lib/src/gutter/error.dart

```diff
@@ -1,14 +1,173 @@
+import '../analyzer/api/models/issue.dart';
+import '../analyzer/api/models/issue_type.dart';
+import '../code_theme/code_theme.dart';
+import '../code_theme/code_theme_data.dart';
 
-class GutterErrorWidget extends StatelessWidget {
-  const GutterErrorWidget();
+const errorIcon = Icon(
+  Icons.cancel,
+  color: Colors.red,
+  size: 16,
+);
+
+const warningIcon = Icon(
+  Icons.warning,
+  color: Colors.yellow,
+  size: 16,
+);
+
+const infoIcon = Icon(
+  Icons.info,
+  color: Colors.white,
+  size: 16,
+);
+
+class GutterErrorWidget extends StatefulWidget {
+  final Issue issue;
+  final TextStyle style;
+
+  const GutterErrorWidget(
+    this.issue, {
+    required this.style,
+  });
+
+  @override
+  State<GutterErrorWidget> createState() => _GutterErrorWidgetState();
+}
+
+class _GutterErrorWidgetState extends State<GutterErrorWidget> {
+  OverlayEntry? entry;
+  bool enteredPopup = false;
+  bool showErrorDetails = false;
 
   @override
   Widget build(BuildContext context) {
-    return const Icon(
-      Icons.cancel,
-      color: Colors.red,
-      size: 16,
+    Icon icon;
+    switch (widget.issue.type) {
+      case IssueType.error:
+        icon = errorIcon;
+        break;
+      case IssueType.warning:
+        icon = warningIcon;
+        break;
+      case IssueType.info:
+        icon = infoIcon;
+        break;
+    }
+
+    final theme = CodeTheme.of(context) ??
+        CodeThemeData(
+          styles: monokaiSublimeTheme,
+        );
+    final backgroundColor = theme.styles['root']?.backgroundColor;
+    final color = theme.styles['root']?.color;
+    final style = widget.style.copyWith(
+      fontSize: 14,
+      color: color,
+      backgroundColor: backgroundColor,
+      fontStyle: FontStyle.normal,
+    );
+    return MouseRegion(
+      onEnter: (event) {
+        setState(() {
+          showErrorDetails = true;
+          enteredPopup = false;
+          if (entry != null) {
+            return;
+          }
+          entry = getErrorPopup(
+            widget.issue,
+            offset: event.position.translate(-5, -5),
+            style: style,
+          );
+          final overlay = Overlay.of(context);
+          overlay?.insert(entry!);
+          overlay?.build(context);
+        });
+      },
+      onExit: (event) {
+        Future.delayed(
+          const Duration(milliseconds: 50),
+          () {
+            setState(() {
+              showErrorDetails = false;
+              if (!enteredPopup) {
+                entry?.remove();
+                entry = null;
+              }
+            });
+          },
+        );
+      },
+      child: icon,
+    );
+  }
+
+  OverlayEntry getErrorPopup(
+    Issue issue, {
+    required Offset offset,
+    required TextStyle style,
+  }) {
+    final renderBox = context.findRenderObject() as RenderBox?;
+    final width = renderBox?.size.width ?? 16;
+    final newOffset = renderBox?.localToGlobal(Offset.zero) ?? offset;
```

Comment:
Return null if renderBox == null.

Reply:
I think it is more convenient to return empty overlay entry rather than null.